File: src/jjnguy/jdapper/JDapper.java
package jjnguy.jdapper;

import java.lang.reflect.Field;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

public class JDapper {
    private final Connection conn;

    public JDapper(Connection conn) {
        this.conn = conn;
    }

    public <T> List<T> query(String sql, Class<T> type, Object... params) throws Exception {
        List<T> results = new ArrayList<>();

        Field[] fields = type.getDeclaredFields();
        for (Field field : fields) {
            field.setAccessible(true);
        }

        PreparedStatement statement = conn.prepareStatement(sql);
        for (int i = 0; i < params.length; i++) {
            statement.setObject(i + 1, params[i]);
        }

        ResultSet rs = statement.executeQuery();

        while (rs.next()) {
            T obj = type.getConstructor().newInstance();

            for (Field field : fields) {
                Object data = rs.getObject(field.getName());
                field.set(obj, data);
            }

            results.add(obj);
        }

        return results;
    }

    private static final String RAW_INSERT = "INSERT INTO %s (%s) VALUES (%s);";

    public void insert(String tableName, Object data) throws Exception {
        Class<?> type = data.getClass();
        Field[] fields = type.getFields();

        StringBuilder columnNames = new StringBuilder();
        StringBuilder values = new StringBuilder();
        for (Field field : fields) {
            columnNames.append(",").append(field.getName());
            values.append(",?");
        }
        columnNames = new StringBuilder(columnNames.substring(1));
        values = new StringBuilder(values.substring(1));

        String sql = String.format(RAW_INSERT, tableName, columnNames.toString(), values.toString());

        PreparedStatement statement = conn.prepareStatement(sql);

        for (int i = 0; i < fields.length; i++) {
            statement.setObject(i + 1, fields[i].get(data));
        }

        statement.execute();
    }
}

File: src/jjnguy/jdapper/JDapper.kt
package jjnguy.jdapper

import java.lang.reflect.Field
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.util.ArrayList

class JDapper(private val conn: Connection) {

  fun <T> query(sql: String, type: Class<T>, vararg params: Any): List<T> {
    val results = ArrayList<T>()

    val fields = type.declaredFields
    for (field in fields) {
      field.isAccessible = true
    }

    val statement = conn.prepareStatement(sql)
    for (i in params.indices) {
      statement.setObject(i + 1, params[i])
    }

    val rs = statement.executeQuery()

    while (rs.next()) {
      val obj = type.getConstructor().newInstance()

      for (field in fields) {
        val data = rs.getObject(field.name)
        field.set(obj, data)
      }

      results.add(obj)
    }

    return results
  }

  fun insert(tableName: String, data: Any) {
    val type = data.javaClass
    val fields = type.fields

    var columnNames = StringBuilder()
    var values = StringBuilder()
    for (field in fields) {
      columnNames.append(",").append(field.name)
      values.append(",?")
    }
    columnNames = StringBuilder(columnNames.substring(1))
    values = StringBuilder(values.substring(1))

    val sql = String.format(RAW_INSERT, tableName, columnNames.toString(), values.toString())

    val statement = conn.prepareStatement(sql)

    for (i in fields.indices) {
      statement.setObject(i + 1, fields[i].get(data))
    }

    statement.execute()
  }

  companion object {

    private val RAW_INSERT = "INSERT INTO %s (%s) VALUES (%s);"
  }
}
